### ISSUE TITLE
support AWS ASG for workers

### ISSUE THREAD

**Issue by rimusz** [juju-solutions/bundle-canonical-kubernetes#123]
Would be nice to support AWS ASG for workers, so workers can be provisioned automaticly

**Comment by rimusz** [comment]
this one should be useful https://github.com/kubernetes/contrib/tree/master/cluster-autoscaler/cloudprovider/aws

**Comment by marcoceppi** [comment]
So, I've been thinking a lot about this. As a result I want to change the title and scope of this feature request. ASG is slightly problematic for a few reasons:

* It's an AWS only system/process
* No direct/easy way to duplicate a running Juju instance

The first one is the one I'm most concerned with. The thing we strive for with Canonical Distribution of Kubernetes (CDK) is that it remains cloud agnostic. For your case, that's not as appealing, but we have many users on bare metal, vmware, open stack, and other clouds who would be interested in the same primitives.

So, I'd like to propose a compromise. This feature, to me, is really "I need autoscaling for workers on my deployment". Juju has an API that the CLI/GUI uses to manipulate the model. I'd like to see this feature as us creating a cluster-autoscaler plugin for kubernetes that uses Juju as a backend.

The workflow would look as such:

- Leverage python-libjuju library for async processing of events in the model for kubernetes-worker
- Use either pykube or kube python packages for communication with Kubernetes
- Translate existing guidelines for consumption -> scale
- Add the following charm configuration to kubernetes-worker:
  - `key: type (default value) # Description`
  - `autoscaling`: `bool` (`False`)
  - `autoscaling-key`: `String` (`None`) # Juju register key to add write access to current model
  - `autoscalling-min`: `Integer` (`2`) # Minimum number of workers
  - `autoscaling-max`: `Integer` (`5`) # Maximum number of workers based on load
- Autoscalling will be configuration on each worker, this allows control of each worker cluster and scaling

Questions I have outstanding:
* Does this need to be delivered as a docker container like the AWS autoscaler is - or could it just run on the worker alongside kubelet
* What happens when an operator scales workers < `autoscaling-min` || > `autoscalling-max`?
* How do we handle faults in the model? IE: Should we yank and replace units in error states?

This will bring autoscaling to every single place CDK runs from bare metal to private/public clouds. If this sounds satisfactory to you @rimusz I'll update the feature request to reflect this start to schedule this work in.

Feedback appreciated.

**Comment by wwwtyro** [comment]
To your questions:

1) I don't immediately see any requirement for nor advantage to delivering this in a container. I am curious where the scaling logic would live & run. k8s master leader?

2) Yeah, that's tricky. At first blush, I'm inclined to require the user to disable autoscaling if they don't want us to immediately overwrite their k8s-worker unit count.

3) Are we talking about k8s node faults, or juju unit faults, here?

I like this plan. :+1:

**Comment by marcoceppi** [comment]
@wwwtyro I should have added I have strong opinions about each of these, but didn't want to make my reply longer than it already was.

> 1) I don't immediately see any requirement for nor advantage to delivering this in a container. I am curious where the scaling logic would live & run. k8s master leader?

I imagine this being installed on each unit but only running on the leader for each worker. While it makes sense to live on master, it does not allow granular control over each deployed group of workers. So in the case of having two worker clusters, say `workers-alpha` and `workers-bravo` where they are tagged as such and different workloads/namespaces live on each. Different rules for autoscalling might apply to each application and not to all workers as a whole

> 2) Yeah, that's tricky. At first blush, I'm inclined to require the user to disable autoscaling if they don't want us to immediately overwrite their k8s-worker unit count.

I agree, to an extent. I think we should treat this as an effective operating range. For example, if the number of units is less than autoscaling-min leader status message would indicate autoscalling: disabled, or something similar and would elect not to bring the cluster up to the minimum units. The same for the inverse with greater than max.

Alternatively, we could provide another configuration option:
- `autoscaling-reconsile`: `bool` (`False`) # Maintain worker unit count within defined range

Which allows for either behavior

> 3) Are we talking about k8s node faults, or juju unit faults, here?

In this case, I'm talking about juju unit faults, but as health checks get added to Juju, also k8s node failures. The immediate questions was framed for things like agent-status: down where a machine goes away. Should autoscalling also correct for cloud failures? In the future, if we have an unhealth k8s machine (for whatever reason), the same question applies.

I will have to dig in to how juju handles agent-status: down for leader units. Do these incur a re-election? If not we may need to re-think our leader-only approach.

**Comment by chuckbutler** [comment]
I can say that historically using leader-only things can be quite prickly if you're not prepared for the hassle of funneling every. single. operation. through. the. leader.

And when that leader goes away, so long as your settings are already established in leader data, you can pick up and restore from where you left off. Even mutate the state a bit. 

But leaders do complicate comprehension of the model at first glance. Its not obvious to anyone who is not familiar with the code that leaders have special roles in charms that utilize this. See the long history of Etcd quorem work, rework, reworking the rework, and finally nailing it down.

Just food for thought as we approach this. There are rough corners there that will require sanding down.

**Comment by marcoceppi** [comment]
Yeah, I'm not really sure how to make sure we don't have /every/ worker spamming the juju api and scaling as it monitors. To me, leadership seems the way, since this isn't an algorithm to run, but some stateless application that churns through live data.

This is why I'm curious if having it in a Docker container running on k8s resolves the issue of trying to figure out how to deal with leadership stuff. All good food for thought if we take this path for autoscaling

**Comment by rimusz** [comment]
hey guys any update on this issue?

**Comment by marcoceppi** [comment]
We're looking into this, I'm starting a POC this week and I'll update you next week

**Comment by rimusz** [comment]
any update on this one?

**Comment by chuckbutler** [comment]
There has been and we are excited to show you at the summit.

**Comment by rimusz** [comment]
Cool 👍 looking forward to it

**Comment by rimusz** [comment]
After talking with @tvansteenburgh the worker scaler could run in a k8s pod, talk to juju API and scale up/down worker that way.

**Comment by simonklb** [comment]
@rimusz and that is almost exactly what we at Elastisys has created (we're running it as a Juju charm and not a k8s pod)! https://jujucharms.com/u/elastisys/charmscaler

I'm currently putting together a bundle with k8s and our autoscaler which will scale the worker nodes on CPU usage. I'll make sure to update this thread when I've pushed it to the charmstore.

You should be fine adding it to your current k8s deployment though. If you have time trying it out, let me know if you have any questions or feature requests!

**Comment by rimusz** [comment]
@simonklb thanks. 
keep me updated with the k8s/autoscaler bundle progress :)

**Comment by simonklb** [comment]
Sorry that I haven't come around to push it until now, the last couple of weeks have been hectic. Here is the first draft: https://github.com/elastisys/bundle-autoscaled-kubernetes/

@rimusz feel free to try it out and let me know if there are any issues or questions!
Should be deployable with `juju deploy cs:~elastisys/bundle/autoscaled-kubernetes-0`
